## Section 1: Warm-up Questions

**Question 1. Element-wise text concatenation** 

Suppose you are given two lists, with their elements at the corresponding position representing the first name and the last name of the same person:

```python
first_names = ['Alice', 'Bob', 'Carol', 'David']
last_names = ['Alligator', 'Bear', 'Chimpanzee', 'Deer']
```

Write a lambda expression, name it `concat_names`, and call it with the two lists above to form a list of full names. The expected output is as follows:


```python
['Alice Alligator', 'Bob Bear', 'Carol Chimpanzee', 'David Deer']
```

In [ ]:
first_names = ['Alice','Bob','Carol','David']
last_names = ['Alligator','Bear','Chimpanzee','Deer']

# Write your code below
concat_names = lambda a,b : [a[i] + ' ' + b[i] for i in range(len(a))]
#or cancat_names = lambda a,b : x + ' ' + y for x, y in zip(a,b)
print(concat_names(first_names, last_names))

['Alice Alligator', 'Bob Bear', 'Carol Chimpanzee', 'David Deer']


---

**Question 2. Use of `lambda`s as function arguments** 

Given a nested list representing gradebooks of different courses:

```python
gradebooks = [[['Troy', 92], ['Alice', 95]], [['James', 89], ['Charles', 100], ['Bryn', 59]]]
```

- Using the builtin `sorted()` function, write code to sort courses by course mean. The expected output is
```python
[[['James', 89], ['Charles', 100], ['Bryn', 59]], [['Troy', 92], ['Alice', 95]]]
```
- Using the builtin `sorted()` function, write code to sort students of each course by score in descending order. The expected output is
```python
[[['Alice', 95], ['Troy', 92]], [['Charles', 100], ['James', 89], ['Bryn', 59]]]
```

In [ ]:
gradebooks = [[['Troy', 92], ['Alice', 95]], [['James', 89], ['Charles', 100], ['Bryn', 59]]]

# Write your code below
task1 = sorted(gradebooks, key = lambda course: sum([score for student, score in course])/ len(course))
print(task1)
task2 = [sorted(course, key = lambda x : x [1], reverse = True) for course in gradebooks]
print(task2)

[[['James', 89], ['Charles', 100], ['Bryn', 59]], [['Troy', 92], ['Alice', 95]]]
[[['Alice', 95], ['Troy', 92]], [['Charles', 100], ['James', 89], ['Bryn', 59]]]


---

**Question 3. Namespaces and Inheritance Search**

Suppose that after running part of a program, the following objects and namespaces are created:

<img src="https://drive.google.com/uc?export=download&id=1fNqROXxfsU_uzLNjEoT6P4UAhM9bVV1M" width=500/>

Note that there are 3 variables defined in the global namespace: `n`, `A`, and `a`.

The code for the `A.foo` function is:

```python
def foo(self):
    return n + self.n
```
The code for the `A.bar` function is:
```python
def bar(self):
    return self.n + A.n + a.n + n
```    

And the code for the `A.__init__` function is:

```python
def __init__(self, x):
    self.n = x
```        

What values are printed if the code below is executed, starting from the state represented by the diagram above?

- `a.foo()`
Answer: 12
- `a.bar()`
Answer: 20

In [ ]:
n = 7
class A:
  n = 3
  def foo(self):
    return n + self.n
  def bar(self):
    return self.n + A.n + a.n + n
  def __init__(self ,x):
    self.n = x
a = A(5)
a.foo(), a.bar()

(12, 20)


## Section 2


For each question below, please write the corresponding code and then run the code to show the output. 






**Question 1**

Many programming languages implement a function called `sum()`, which returns the sum of an arbitrary sequence of numbers.

Define a function `mySum` that allows us to sum up any number of numeric values. For example:

- We should be able to call `mySum(1, 12, 89, 5)` with 4 arguments or `mySum(*range(100))` with 100 arguments. 
- Calling `mySum()` with no arguments returns zero.

The built-in function `sum()` is not allowed to use in the function body.

In [1]:
# Write your function definition below
def mySum(*listX):
  if not listX:
    return 0
  total = 0
  for item in listX:
    total += item
  return total

In [2]:
mySum(1, 12, 89, 5)

107

In [3]:
mySum(*range(100))

4950

In [4]:
mySum()

0

---

**Question 2**

Parallel processing of multiple sequences is a powerful functionality for data analysis. 

Define a function `mySumParallel(seqs)` that applies `mySum()` defined in **Question 1** to an arbitrary collection of sequences of numeric values in parallel. This function returns a list of sums of all sequences in the passed-in collection.

For example, calling `mySumParallel(collection)` where `collection = [[1, 12, 89, 5], range(100), range(2, 9, 3)]` returns `[107, 4950, 15]`.



In [5]:
collection = [[1, 12, 89, 5], range(100), range(2, 9, 3)]

# Write your function definition below
def mySumParallel(seqs):
  result = []
  for elem in seqs:
    result.append(mySum(*elem))
  return result



In [6]:
mySumParallel(collection)

[107, 4950, 15]

---

**Question 3**

Further define a function `myParallel(aggfunc, seqs)` that allows any aggregation function to be applied to an arbitrary collection of sequences of numeric values.

For example, calling `myParallel(mySum, collection)` yields `[107, 4950, 15]`, while calling `myParallel(max, collection)` yields `[89, 99, 8]`.

In [9]:
# Write your function definition below
def myParallel(aggfunc, seqs):
  result = []
  for elem in seqs:
    result.append(aggfunc(*elem))
  return result





In [10]:
print("myParallel(mySum, collection):{}".format(myParallel(mySum, collection)))

myParallel(mySum, collection):[107, 4950, 15]


In [11]:
print("myParallel(max, collection):{}".format(myParallel(max, collection)))

myParallel(max, collection):[89, 99, 8]


---

**Question 4. Matrix Construction** 


In mathematics, a matrix is a 2-dimensional (tabular) array of numbers or symbols arranged in rows and columns. The following is the notation of a 3 by 2 matrix (3 rows and 2 columns) containing the numbers from 1 to 6:

\begin{align}
\\
matrixA & =
 \begin{pmatrix}
  1 & 2  \\
  3 & 4  \\
  5 & 6 \\
 \end{pmatrix}
\end{align}

In Python, we can represent this 2-dimensional array with the nested list

```python
matrixA = [[1, 2], [3, 4], [5, 6]]
```
where the number of first-level elements represent the number of rows and the number of second-level elements represent the number of columns.

 
Write a function `create_matrix` that takes a sequence of integers and two additional numbers that specify the shape of the matrix (i.e., the number of rows and columns, respectively) to create a nested list representing the matrix filled by these integers row-wise.
  
  For example, calling `create_matrix(*range(4, 19), rows=3, cols=5)` returns a nested list representing the following 3 (rows) by 5 (columns) matrix:
\begin{align}
\\
 \begin{pmatrix}
  4 & 5 & 6 & 7 & 8 \\
  9 & 10 & 11 & 12 & 13  \\
  14 & 15 & 16 & 17 & 18 \\
 \end{pmatrix}
\end{align}    
  Also include exception handling so that the function prints an error message if the number of intergers does not match the product of the number of rows and that of columns. For example, calling `create_matrix(2, 3, 14, 15, rows=3, cols=2)` prints 
 ```
The number of elememnts does not match the shape of the matrix.
``` 


In [ ]:
# Write your function definition below
def create_matrix(*seqs, rows, cols):
  if rows * cols != len(seqs):
    print("The number of elements does not match the shape of the matrix.")
    return
  else:
    return [[seqs[b * cols + a] for a in range(cols)] for b in range(rows)]
create_matrix(*range(4, 19), rows=3, cols=5)  

[[4, 5, 6, 7, 8], [9, 10, 11, 12, 13], [14, 15, 16, 17, 18]]

---


 
**Question 5**

Suppose a music streaming service uses two lists to maintain a user's favorite songs and the corresponding ratings, respectively:

```python
songs = ['Back to Black', 'Poker Face', 'Yellow', 'Lolipop', 'All Too Well', 'Delicate', 'Moves Like Jagger']
ratings = [9.5, 8, 9, 8, 10, 7.5, 8]
```

Write a **signle expression** to create a playlist, which orders the user's favorite songs based on their ratings. And if a tie is observed, include the lastest one first (the one closest to the end of the list).

The expected output should look like the following:

```python
['All Too Well', 'Back to Black', 'Yellow', 'Moves Like Jagger', 'Lolipop', 'Poker Face', 'Delicate']
```

- Hint: use the slicing notation `[-1::-1]` to reverse the order of elements in `songs` and `ratings`.

In [ ]:
songs = ['Back to Black', 'Poker Face', 'Yellow', 'Lolipop', 'All Too Well', 'Delicate', 'Moves Like Jagger']
ratings = [9.5, 8, 9, 8, 10, 7.5, 8]
playlist = [name for name, score in sorted(zip(songs, ratings), key = lambda x: x[1])][-1::-1] 
print(playlist)
# Write your code below




['All Too Well', 'Back to Black', 'Yellow', 'Moves Like Jagger', 'Lolipop', 'Poker Face', 'Delicate']


---

**Question 6. Use of `ipywidgets` (bonus question)** 


The Jupyter notebook and its IPython kernel support the creation of interactive HTML widgets (i.e., elements of graphical user interface). These widgets allow us to control the code and the data by responding to events and invoking specified handlers.

In order to incorporate widgets in the notebook, we have to import the `ipywidgets` module first, as shown below:


In [ ]:
import ipywidgets

To add a slider that selects an integer in a range, we create an `IntSlider` instance by passing an initial value, the minimum and maximum values, the interval size (step), and the description to the `IntSlider` class's constructor:

In [ ]:
help(ipywidgets.IntSlider.__init__)

Help on function __init__ in module ipywidgets.widgets.widget_int:

__init__(self, value=None, min=None, max=None, step=None, **kwargs)
    Parameters
    ----------
    value: integer
        The initial value.
    min: integer
        The lower limit for the value.
    max: integer
        The upper limit for the value.
    step: integer
        The step between allowed values.



In [ ]:
slider = ipywidgets.IntSlider(
    min=0,
    max=10,
    step=1,
    description='input',
    value=3
)

The IPython.display.display function can be used to render the slider:

In [ ]:
from IPython.display import display
display(slider)

IntSlider(value=3, description='input', max=10)

If we want to use `slider` to pick an input for producing a certain output interactively, we can use the `interactive_output` function that connects the widget controls to a function that renders the output:


In [ ]:
help(ipywidgets.interactive_output)

Help on function interactive_output in module ipywidgets.widgets.interaction:

interactive_output(f, controls)
    Connect widget controls to a function.
    
    This function does not generate a user interface for the widgets (unlike `interact`).
    This enables customisation of the widget user interface layout.
    The user interface layout must be defined and displayed manually.



In [ ]:
input_a = ipywidgets.IntSlider(max=10, description='a')     
input_b = ipywidgets.IntSlider(max=10, description='b')

def f(a, b):
    if a < b: print('a={} is less than b={}'.format(a, b))    
    else: print('a={} is greater than or equal to b={}'.format(a, b))

# The controls arguement specifies how inputs from the two sliders are passed into the call to the function f
output = ipywidgets.interactive_output(f, {'a': input_a, 'b': input_b})    

Call the `IPython.display.display` function to render the 3 widget elements simultaneously:

In [ ]:
display(input_a, input_b, output)

IntSlider(value=0, description='a', max=10)

IntSlider(value=0, description='b', max=10)

Output()

Revise the code above to create a widget that takes inputs from two sliders and calculates the product of integers from 0~100 (both inclusive).

In [ ]:
# Write your code below
input_a = ipywidgets.IntSlider(max=100, description='a')     
input_b = ipywidgets.IntSlider(max=100, description='b')

def f(a, b):
    print("Product result: {}".format(a*b))

# The controls arguement specifies how inputs from the two sliders are passed into the call to the function f
output = ipywidgets.interactive_output(f, {'a': input_a, 'b': input_b})   
display(input_a, input_b, output)

IntSlider(value=0, description='a')

IntSlider(value=0, description='b')

Output()

You can refer to this [website](https://ipywidgets.readthedocs.io/en/latest/index.html) for more information on Jupyter widgets.